magenta-lib/src/main/scala/magenta/tasks/ChangeSetTasks.scala

```diff
@@ -147,6 +168,19 @@ class ExecuteChangeSetTask(
 
         val request = ExecuteChangeSetRequest.builder().changeSetName(changeSetName).stackName(stackName).build()
         cfnClient.executeChangeSet(request)
+
+        new CloudFormationStackEventPoller(stackName, cfnClient, resources, stopFlag)
```

Comment:
I'm not sure why `CloudFormationStackEventPoller` needs to be a class here. It only exposes a single method,`check`, and only that method relies on the class constructor parameters/state as far as I can tell.

In general, I would say use classes for data, and objects with methods (which is a form of namespacing really) for behaviour. There are exceptions, but typically this is simplest and best.

Reply:
> I'm not sure why `CloudFormationStackEventPoller` needs to be a class here.

As `check` is called recursively, with only `lastSeenEvent` varying between calls, it felt unnecessary to pass the cfn client etc each time. My Scala is quite stale atm - pairing/suggestions welcomed!

Reply:
@akash1810 here's one way to do it (below). Note: an inner loop is used, though you could just recur here without this. A class (with constructor params) is only serving to fragment the true signature of `check` here (as none of the other class members use the constructor params). It's better, I think, to be explicit and accept that the function signature is a wide one.

```
object CloudFormationStackEventPoller extends Loggable {
  private[this] def reportEvent(reporter: DeployReporter, e: StackEvent): Unit = {
    reporter.info(s"${e.logicalResourceId} (${e.resourceType}): ${e.resourceStatusAsString}")
    if (e.resourceStatusReason != null) reporter.verbose(e.resourceStatusReason)
  }

  private[this] def isStackEvent(stackName: String)(e: StackEvent): Boolean =
    e.resourceType == "AWS::CloudFormation::Stack" && e.logicalResourceId == stackName

  private[this] def updateStart(stackName: String)(e: StackEvent): Boolean =
    isStackEvent(stackName)(e) && (e.resourceStatusAsString == "UPDATE_IN_PROGRESS" || e.resourceStatusAsString == "CREATE_IN_PROGRESS")

  private[this] def updateComplete(stackName: String)(e: StackEvent): Boolean =
    isStackEvent(stackName)(e) && (e.resourceStatusAsString == "UPDATE_COMPLETE" || e.resourceStatusAsString == "CREATE_COMPLETE")

  private[this] def updateFailed(e: StackEvent): Boolean = {
    val failed = e.resourceStatusAsString.contains("FAILED") || e.resourceStatusAsString.contains("ROLLBACK")
    logger.debug(s"${e.resourceStatusAsString} - failed = $failed")
    failed
  }

  private[this] def fail(reporter: DeployReporter, e: StackEvent): Unit = reporter.fail(
    s"""${e.logicalResourceId}(${e.resourceType}}: ${e.resourceStatusAsString}
       |${e.resourceStatusReason}""".stripMargin)

  def check(
    stackName: String,
    cfnClient: CloudFormationClient,
    resources: DeploymentResources,
    stopFlag: => Boolean,
    lastSeenEvent: Option[StackEvent],
    additionalChecks: () => Boolean = () => true): Unit = {

    def loop(lastSeenEvent: Option[StackEvent], additionalChecks: () => Boolean = () => true): Unit = {
      val result = CloudFormation.describeStackEvents(stackName, cfnClient)
      val events = result.stackEvents.asScala

      lastSeenEvent match {
        case None =>
          events.find(updateStart(stackName)) match {
            case None => resources.reporter.fail(s"No events found at all for stack $stackName")
            case Some(e) =>
              val age = new Duration(new DateTime(e.timestamp().toEpochMilli), new DateTime()).getStandardSeconds
              if (age > 30) {
                resources.reporter.verbose("No recent IN_PROGRESS events found (nothing within last 30 seconds)")
              } else {
                reportEvent(resources.reporter, e)
                check(Some(e), additionalChecks)
              }
          }
        case Some(event) =>
          val newEvents = events.takeWhile(_.timestamp.isAfter(event.timestamp))
          newEvents.reverse.foreach(reportEvent(resources.reporter, _))
          newEvents.filter(updateFailed).foreach(fail(resources.reporter, _))

          val complete = newEvents.exists(updateComplete(stackName)) && additionalChecks()
          if (!complete && !stopFlag) {
            Thread.sleep(5000)
            loop(Some(newEvents.headOption.getOrElse(event)), additionalChecks)
          }
      }

      loop(lastSeenEvent, additionalChecks)
    }
  }
}
```

Reply:
Yeah, that is nicer! I've even started a [refactor](https://github.com/guardian/riff-raff/compare/aa-refactor) that looks almost identical!

> In general, I would say use classes for data, and objects with methods (which is a form of namespacing really) for behaviour. There are exceptions, but typically this is simplest and best.

💯